### PR TITLE
Custom implementation of reciprocal_rank_fusion

### PR DESCRIPTION
**Pull request by x65han**
- add a custom implementation of reciprocal_rank_fusion, so we have control over the implementation details rather than relying on others.

### CODE REVIEW

File: pyserini/search/fusion.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            self._load_qrun(path, k, rrf_by_topics)
+
+        if os.path.exists(output_path):
+            os.remove(output_path)
+
+        with open(output_path, 'w') as f:
+            for topic in rrf_by_topics:
+                rrf_by_topic = rrf_by_topics[topic]
+
+                for index, (doc_id, score) in enumerate(sorted(rrf_by_topic.items(), key=lambda x: x[1], reverse=True)):
```
Comment (lintool) [inline pyserini/search/fusion.py:80]: Add in the _k_ value that trectools also implements?

**Comment by lintool** [comment]
hi @x65han - I was thinking: `reciprocal_rank_fusion` should be a function that takes in multiple hits lists and returns a fused list.

`SimpleFusionSearcher` should be a class in `pyserini.search` that behaves like `SimpleSearcher` but takes multiple indexes.

So something like:

```python
searcher = SimpleFusionSearcher('/path/to/index1/', '/path/to/index2/', ..., method=rrf)
hits = searcher.search('hubble space telescope')

# Print the first 10 hits:
for i in range(0, 10):
    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')
```

And internally, it uses: `reciprocal_rank_fusion`

File: pyserini/search/fusion.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                topic, _, doc_id, rank, _, _ = line.split()
+                if topic not in res:
+                    res[topic] = dict()
+
+                if doc_id not in res[topic]:
+                    res[topic][doc_id] = 0
+
+                res[topic][doc_id] += (1/(k+float(rank)))
+
+    def reciprocal_rank_fusion(self, output_path: str, k: int = 60, tag: str = None):
```
Comment (x65han) [inline pyserini/search/fusion.py:48]: I have the k value passed over here. and used in line 46.

```
res[topic][doc_id] += (1/(k+float(rank)))
```

File: pyserini/search/fusion.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        with open(path, 'r') as f:
+            for line in f:
+                topic, _, doc_id, rank, _, _ = line.split()
+                if topic not in res:
+                    res[topic] = dict()
+
+                if doc_id not in res[topic]:
+                    res[topic][doc_id] = 0
+
+                res[topic][doc_id] += (1/(k+float(rank)))
```
Comment (rodrigonogueira4) [inline pyserini/search/fusion.py:46]: nit: no need to convert to float in python 3

**Comment by x65han** [comment]
Abandoning after code structure change.

**Pull request closed without merging**